File: src-lang/xt/db/node/proxy_supabase.clj
(ns xt.db.node.proxy-supabase
  (:require [hara.lang :as l]))

(l/script :xtalk
  {:require [[xt.lang.spec-base :as xt]
             [xt.lang.spec-promise :as promise]
             [xt.lang.common-data :as xtd]
             [xt.substrate :as substrate]
             [xt.substrate.page-proxy :as page-proxy]
             [xt.db.node.proxy-util :as proxy-util]]})

(def.xt ACTIONS
  ["@xt.supabase/sign-up"
   "@xt.supabase/sign-in"
   "@xt.supabase/sign-out"
   "@xt.supabase/refresh"
   "@xt.supabase/signed-in?"
   "@xt.supabase/current-session"
   "@xt.supabase/rpc-call"
   "@xt.supabase/query-table"
   "@xt.supabase/health"
   "@xt.supabase/admin-create-user"
   "@xt.supabase/admin-delete-user"
   "@xt.supabase/admin-generate-link"
   "@xt.supabase/admin-get-user"
   "@xt.supabase/admin-list-users"
   "@xt.supabase/admin-update-user"
   "@xt.supabase/authorize"
   "@xt.supabase/callback"
   "@xt.supabase/invite"
   "@xt.supabase/otp"
   "@xt.supabase/recovery"
   "@xt.supabase/settings"
   "@xt.supabase/token-refresh"
   "@xt.supabase/user-get"
   "@xt.supabase/user-info"
   "@xt.supabase/user-put"
   "@xt.supabase/verify-get"
   "@xt.supabase/verify-post"])

(def.xt ATTACH_ACTIONS
  ["@xt.supabase/attach-model"])

(defn.xt attach-forward-handler
  "forwards a Supabase attach action to the server and opens the proxy group"
  {:added "4.1"}
  [space args request node]
  (var page-args (xt/x:second args))
  (var space-id (xtd/get-in page-args ["space_id"]))
  (var group-id (xtd/get-in page-args ["group_id"]))
  (var transport-id (proxy-util/get-transport-id node (xtd/get-in request ["meta"])))
  (page-proxy/group-create-proxy node space-id group-id {} {"transport_id" transport-id})
  (return
   (promise/x:promise-then
    (substrate/request node
                       nil
                       (xt/x:get-key request "action")
                       args
                       {"transport_id" transport-id})
    (fn [status]
      (return
       (promise/x:promise-then
        (page-proxy/group-open-proxy node space-id group-id {"transport_id" transport-id})
        (fn [_]
          (return status))))))))

(defn.xt init-proxy-handlers
  "Registers client-side supabase proxy handlers so that the same substrate
   function ids used server-side can be invoked on a client node and forwarded
   to the server."
  {:added "4.1"}
  [node]
  (xt/for:array [action -/ACTIONS]
    (substrate/register-handler node action proxy-util/request-proxy nil))
  (xt/for:array [action -/ATTACH_ACTIONS]
    (substrate/register-handler node action -/attach-forward-handler nil))
  (return node))

File: packages-gen/js/db/node/proxy-supabase.js
const xtd = require("@xtalk/lang/common-data.js")

const page_proxy = require("@xtalk/substrate/page-proxy.js")

const proxy_util = require("@xtalk/db/node/proxy-util.js")

const substrate = require("@xtalk/substrate/substrate.js")

var ACTIONS = [
  "@xt.supabase/sign-up",
  "@xt.supabase/sign-in",
  "@xt.supabase/sign-out",
  "@xt.supabase/refresh",
  "@xt.supabase/signed-in?",
  "@xt.supabase/current-session",
  "@xt.supabase/rpc-call",
  "@xt.supabase/query-table",
  "@xt.supabase/health",
  "@xt.supabase/admin-create-user",
  "@xt.supabase/admin-delete-user",
  "@xt.supabase/admin-generate-link",
  "@xt.supabase/admin-get-user",
  "@xt.supabase/admin-list-users",
  "@xt.supabase/admin-update-user",
  "@xt.supabase/authorize",
  "@xt.supabase/callback",
  "@xt.supabase/invite",
  "@xt.supabase/otp",
  "@xt.supabase/recovery",
  "@xt.supabase/settings",
  "@xt.supabase/token-refresh",
  "@xt.supabase/user-get",
  "@xt.supabase/user-info",
  "@xt.supabase/user-put",
  "@xt.supabase/verify-get",
  "@xt.supabase/verify-post"
];

var ATTACH_ACTIONS = ["@xt.supabase/attach-model"];

function attach_forward_handler(space,args,request,node){
  let page_args = args[1];
  let space_id = xtd.get_in(page_args,["space_id"]);
  let group_id = xtd.get_in(page_args,["group_id"]);
  let transport_id = proxy_util.get_transport_id(node,xtd.get_in(request,["meta"]));
  page_proxy.group_create_proxy(node,space_id,group_id,{},{"transport_id":transport_id});
  return substrate.request(node,null,request["action"],args,{"transport_id":transport_id}).then(function (status){
    return page_proxy.group_open_proxy(node,space_id,group_id,{"transport_id":transport_id}).then(function (_){
      return status;
    });
  });
}

function init_proxy_handlers(node){
  for(let action of ACTIONS){
    substrate.register_handler(node,action,proxy_util.request_proxy,null);
  };
  for(let action of ATTACH_ACTIONS){
    substrate.register_handler(node,action,attach_forward_handler,null);
  };
  return node;
}

module.exports = {
  ["ACTIONS"]:ACTIONS,
  ["ATTACH_ACTIONS"]:ATTACH_ACTIONS,
  ["attach_forward_handler"]:attach_forward_handler,
  ["init_proxy_handlers"]:init_proxy_handlers
}